### ISSUE TITLE
Improve throttling - update pnpjs - HTTP 429 Error

### ISSUE THREAD

**Issue by wobba** [microsoft-search/pnp-modern-search#3225]
### Discussed in https://github.com/microsoft-search/pnp-modern-search/discussions/3178

<div type='discussions-op-text'>

<sup>Originally posted by **meleila** August  1, 2023</sup>
Hi,
we have a problem with loading the pnp search results web part. Each time a user opens the site for the first time the web part shows the HTTP 429 error. After a reload everything works fine.

Therefore, the error is difficult to reproduce, resetting cache doesnt reproduce the issue again. We can only ask users who never called the site to test.

If we open the issue in the network tab of the dev tool the following error shows: "-1, Microsoft.Office.Server.Search.REST.SearchServiceException"

Modern search version: 4.8
Browser: Edge
</div>

**Comment by gretchunkim** [comment]
@wobba  Received throttle error/issued  for pulling user profiles pictures using SharePoint as data source.  The user was not able visit other SPO while the throttle error clears, in about a good 30 minutes.

**Comment by kasperbolarsen** [comment]
As I understand throttling a single request for user profile data can't by
itself consume the tenants allowance,  but it can of course be the request
that hits the limit.
What are you suggesting should be changed in the solution?

**Comment by gretchunkim** [comment]
it didn't do a single request. It ran the entire 3000 requests (all user profiles in the tenant) at a time. :D

**Comment by kasperbolarsen** [comment]
So the query was like * on the local people source? Or?

**Comment by gretchunkim** [comment]
Exactly. it's SharePoint Online data source, local people source - nothing special.

**Comment by wobba** [comment]
As with any modern page you add custom code to, follow guidance at https://learn.microsoft.com/en-us/microsoft-365/enterprise/modern-web-part-optimization to check what a particular page setup does of API queries.

We cannot prevent how you setup search web parts which adds API queries to the page.

That said, we want to update pnp js which handles retry and pause better. But once you get throttled, you have have to wait. Nothing the web parts can do.

**Comment by wobba** [comment]
Closing — this is Microsoft Graph/SharePoint throttling behavior on profile-picture requests, which is outside the control of these web parts. Please reopen if you have a concrete change to suggest in the solution itself. Thanks!
